timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/NamedMinTimestampLockCollection.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.timelock.lock;
+
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.LoadingCache;
+import java.util.Optional;
+
+final class NamedMinTimestampLockCollection {
```

Comment:
See for what the full API will look like https://github.com/palantir/atlasdb/compare/aa-nmtl-trackers?expand=1.

I will add two methods that are more generic, e.g. `getNamedMinTimestampLock`. But, these methods should refuse anything that isn't `CommitImmutableTimestamp` (and especially anything that is equal to `ImmutableTimestmap`)

Reply:
I don't think I want to settle on a name "COmmitImmutableTimestamp" we're very explicitly making this open to whatever the user wants to do. So let's not add an allowlist in Timelock

What I WOULD like to see is for you to add metrics on how many of these we're holding, and the rough sizes of these collections, could you do that in a FLUP?

Reply:
And, obviously, standardize the metrics for min value etc.

Reply:
> we're very explicitly making this open to whatever the user wants to do.

For what it's worth I'd want us to have control on who uses this feature (or, at least, make it very obvious that unless you know what you're doing you should not use this feature, and we will not support issues where users use this in non-atlasdb team approved ways; @sverma30 for SA). Though yeah controlling this in timelock is a bit heavy-handed as it would force a product dependency on the related product version.